pytensor/tensor/rewriting/linalg.py

```diff
@@ -377,3 +384,94 @@ def local_lift_through_linalg(
                 return [block_diag(*inner_matrices)]
             else:
                 raise NotImplementedError  # pragma: no cover
+
+
+# Det Diag Rewrites
+@register_canonicalize
+@register_stabilize
+@node_rewriter([det])
+def rewrite_det_diag_from_eye_mul(fgraph, node):
+    """
+     This rewrite takes advantage of the fact that for a diagonal matrix, the determinant value is the product of its diagonal elements.
+
+    The presence of a diagonal matrix is detected by inspecting the graph. This rewrite can identify diagonal matrices that arise as the result of elementwise multiplication with an identity matrix. Specialized computation is used to make this rewrite as efficient as possible, depending on whether the multiplication was with a scalar, vector or a matrix.
+
+    Parameters
+    ----------
+    fgraph: FunctionGraph
+        Function graph being optimized
+    node: Apply
+        Node of the function graph to be optimized
+
+    Returns
+    -------
+    list of Variable, optional
+        List of optimized variables, or None if no optimization was performed
+    """
+    # Find if we have Blockwise Op
+    if not (isinstance(node.op, Blockwise) and isinstance(node.op.core_op, Det)):
+        return None
+
+    node_input_op = node.inputs[0]
+    # Check if the op is Elemwise and mul
+    if not (
+        node_input_op.owner is not None
+        and isinstance(node_input_op.owner.op, Elemwise)
+        and isinstance(node_input_op.owner.op.scalar_op, Mul)
+    ):
+        return None
+
+    # Find whether any of the inputs to mul is Eye
+    inputs_to_mul = node_input_op.owner.inputs
+    eye_check = False
+    eye_input = []
+    for mul_input in inputs_to_mul:
+        if isinstance(mul_input.owner.op, Eye):
+            eye_check = True
+            eye_input.append(mul_input)
+            break  # This is so that we only find the first one? Can be changed
+    if not (eye_check):
+        return None
+
+    # Get all non Eye inputs (scalars/matrices/vectors)
+    non_eye_inputs = list(set(inputs_to_mul) - set(eye_input))
+
+    # Dealing with only one other input
+    if len(non_eye_inputs) >= 2:
+        raise NotImplementedError
+
+    # Checking if original x was matrix
+    if not non_eye_inputs[0].owner:
+        # It was a matrix
+        det_val = non_eye_inputs[0].diagonal().prod()
+    else:
+        # Check for scalar (ndim = 0) or vector (ndim = 1)
+        sca_vec_input = non_eye_inputs[0].owner.inputs[0]
+        if sca_vec_input.ndim == 0:
+            det_val = sca_vec_input ** (eye_input[0].shape[0]).astype(config.floatX)
+        else:
+            det_val = sca_vec_input.prod()
```

Comment:
Elemwise operation (like Mul) will always add an `expand_dims` to inputs so they have the same number of dimensions. You can check for `non_eye_inputs[0].type.broadcastable` to know whether the input is being broadcasted or not, regardless of if it's a root variable or coming out of another operation.

`scalar  * eye`, will be converted into `expand_dims(scalar, axis=(0, 1)) * eye`, `vector * eye`  will be converted into `expand_dims(vector, axis=(0)) * eye`. You also have to think about other edge cases, what if it was a row/col matrix being multiplied by eye? What if either eye or the multiplication are more than 2D (I don't know if eye is a Blockwise and can already be more than 2D, but the other input certainly can)

Reply:
i added a test for row and column matrices separately and its working right now.
im a lil confused by what you want me to change.
instead of the `ndim` check, should i use the `.type.broadcastable` and use the axes to get the typeof input?

Reply:
For instance you have this comment above
```python
# Checking if original x was matrix
if not non_eye_inputs[0].owner:
```

If you want to check if an input is a matrix, which I assume you mean a 2d tensor that is not being broadcasted (because the input will always be ndim=2 after the multiplication), you would check `non_eye_inputs[0].type.broadcastable == (False, False)`. Instead a scalar that is being broadcasted would have a pattern `(True, True)`, and a row/col vector would have one True and one False.

You don't need to check if the input has any owner (meaning it comes out of another operation) or anything like that, because it' doesn't matter. All you need to know is whether you had a full rank matrix, or some other object that is being broadcasted in the multiplication with eye.

Does that clarify it?

Reply:
ohhh that makes sense. yea it was kind of annoying to separately check for matrix that way

Reply:
if i use this way to check for the type of input, it throws an error for row and column matrices. how should i check differently for them?

Reply:
this is happening because when its checking for them with the `.type.broadcastable`, it treats them as a vector even though they are matrices